src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj

```diff
@@ -1,64 +1,124 @@
 (ns puppetlabs.trapperkeeper.services.scheduler.scheduler-core
-  (:require [overtone.at-at :as at-at]
-            [clojure.tools.logging :as log]
-            [puppetlabs.i18n.core :as i18n]))
+  (:require [clojure.tools.logging :as log]
+            [puppetlabs.i18n.core :as i18n]
+            [puppetlabs.kitchensink.core :as ks])
+  (:import (org.quartz.impl.matchers GroupMatcher)
+           (org.quartz.impl StdSchedulerFactory SchedulerRepository)
+           (org.quartz JobBuilder SimpleScheduleBuilder TriggerBuilder DateBuilder
+                       DateBuilder$IntervalUnit Scheduler JobKey SchedulerException JobDataMap)
+           (org.quartz.utils Key)))
 
-(defn create-pool
+(def shutdown-timeout-sec 30)
+(def instance-count (atom 0))
+
+(defn- make-instance-name
+  [count]
+  (str "tk-scheduler-" count))
+
+(defn create-scheduler
   "Creates and returns a thread pool which can be used for scheduling jobs."
   []
-  (at-at/mk-pool))
+  ;; without the following property set, quartz does a version check automatically
+  (System/setProperty "org.quartz.scheduler.skipUpdateCheck" "true")
+  (let [next-instance-count (swap! instance-count inc)
+        sr (SchedulerRepository/getInstance)]
+    ; force each scheduler instance to have a unique name.  This is important on HUP as the previous scheduler
+    ; jobs may not have finished which marks it as "not shutdown", and the normal behavior of the std scheduler
+    ; is to lookup the scheduler in the SchedulerRepository and reuse the existing one first
+    (System/setProperty "org.quartz.scheduler.instanceName" (make-instance-name next-instance-count))
+    ; explicitly remove any old instance from the scheduler repository to prevent leaks
+    (.remove sr (make-instance-name (- next-instance-count 1)))
```

Comment:
I'm grokking the pros and cons to doing this here vs in stop-all-jobs, which, since it shutdown the scheduler instance is where I'd figure I'd see it removed from the registry.

Reply:
Good point.